### ISSUE TITLE
RFC: Allow developers to use a special character in combination with a new serializer to prevent CF from serializing strings containing only numeric data as numeric values

### ISSUE THREAD

**Issue by atuttle** [atuttle/Taffy#321]
**Problem:** Some text fields will sometimes contain only numeric data, and when CF's implementation of SerializeJson sees this, it attempts to be helpful and convert that to a numeric. E.g. when a phone number doesn't have any hyphens or other punctuation:

```js
[{
  "phone": "555-867-5309"
},
{
  "phone": 5558675309
}]
```

**Proposed Solution:** Establish a convention for a non-printable special character that Taffy will automatically strip from all data after json serialization. Probably via a new opt-in serializer, something like this:

```js
component extends="taffy.core.nativeJsonSerializer" {
  function getAsJson() taffy:mime="application/json;text/json" taffy:default="true" {
    var serialized = super.getAsJson();
    var clean = serialized.replace( chr(2), "", "ALL" );
    return clean;
  }
}
```

`chr(2)` is the special non-printable character "start-text" (STX), and has been shown to force CF to treat a variable as a string when passed through SerializeJson, E.g:

```js
{
  "phone": "{STX}5558675309"
}
```

With this opt-in serializer in use, you could inject a `chr(2)` into any variable to force it to render as a string, e.g.:

```js
return rep(queryToArray(myQuery, function(row){
  row.phone = chr(2) & row.phone;
  return row;
}));
```

After the serializer has done its job, the final output will be (a string of numerics, not a numeric value):

```js
{
  "phone": "5558675309"
}
```

---
This approach does add a little bit more overhead in both rendering your query to an array, and then stripping the special characters from your serialized json, but it is a guaranteed outcome, and one that some people may prefer over a field that is sometimes numeric, sometimes a string, depending on its contents.

**Comment by dswitzer** [comment]
If you're going to do something like this, would seem best to have a helper method for transforming the values:

JsonSerializer.toNumeric(row.phone);
JsonSerializer.toString();

One other thought, seems to be I recall if you transform values to Java natives, the native serializer seems to handle things okay. I wonder if you could just have the helper functions cast values to Java natives.

**Comment by mjclemente** [comment]
I'm for this approach. I'll take the performance hit in return for a guaranteed outcome. Thanks for the fantastic project @atuttle .

**Comment by atuttle** [comment]
@dswitzer good point. I'd make them methods available inside resources; something like this?

```js
return rep(queryToArray(myQuery, function(row){
  row.phone = forceString(row.phone);
  return row;
}));
```

**Comment by dswitzer** [comment]
@atuttle:

Personally, I'd namespace everything, to reduce possible collisions:

```
return rep(queryToArray(myQuery, function(row){
  row.phone = Serialize.toString(row.phone);
  return row;
}));
```
Or maybe "Encoder" would be better than "Serialize". Either way, load the methods into a single variable and expose that variable to the resources.

**Comment by atuttle** [comment]
`toString` is likely to be a problematic name because it's a very common method on Java objects and even available in some CFML contexts... But I see your point. I'll consider it.

**Comment by pfreitag** [comment]
Just wanted to add that problems can also occur when ACF thinks it sees a boolean, for example if you have:

`name = {first="Dr.", last="No"};`

In ACF it will serialize as:

`{ "first": "Dr.", "last":false}`

I would think your solution would work well for this case as well.

**Comment by atuttle** [comment]
I decided to build it into the default serialization pipeline rather than making it an opt-in feature. This will affect anyone who is returning `STX` characters intentionally because it will start stripping them out automatically... but I have to imagine that is such a remote possibility that it is probably not worth worrying about.

It's implemented as `encode.string()`, so to update my example from before:

```js
return rep(queryToArray(myQuery, function(row){
  row.phone = encode.string(row.phone);
  return row;
}));
```

This is available in master (not in a tagged release) right now. I've only tested it on ACF11, but I tried to be mindful of our support back to ACF8 and Lucee. If anyone could test on other platforms/versions, that would be 👍 

I've updated the most basic example (`/examples/api/`) to have two new data elements for the artMember resource. If you get this response, then it would be considered working as expected:

![image](https://cloud.githubusercontent.com/assets/46990/17179071/e20045de-53e5-11e6-8075-1a2a7274f483.png)

**Comment by ryanguill** [comment]
Are the labels up to date on this issue - it looks like you've got at least a start on it - is there more to do?

**Comment by atuttle** [comment]
It's functioning in master but not yet documented or part of an official release. I'm using it in production. :)

**Comment by anteroortiz** [comment]
FYI: On Coldfusion 11 Update 11, a value of "1" now returns "\u00021".

It works properly on update 10.

**Comment by atuttle** [comment]
@anteroortiz can you be more specific? I've been using this change in production since it was made without issue. I don't think I've tried it on just the number 1 though. Is it just that number or all numerics that are affected?

**Comment by anteroortiz** [comment]
Yes, I'm basically using the number 1 as an example for all integers given that they're easy values to test in terms of string conversion. In terms of the JSON output, a value of 1 was treated like this:

a. Do nothing => 1
b. encode() on CF11 Update 10 => "1"
c. encode() on CF11 Update 11 =>  "\u00021"

**Comment by anteroortiz** [comment]
Yes, I'm seeing it for all intergers. And for values such as 2.5. Interesting that encode.string("hello") also gives me: "\u0002hello"

If you're not experiencing it on Update 11, then maybe I have something wrong between the home made version (before added to repo) versus the one in the library. But then it'd be odd that everything works on Update 10. Only update 11 is behaving this way.

**Comment by atuttle** [comment]
@anteroortiz I am able to reproduce this. The reason we haven't seen this in production yet is because we haven't moved production to update 11 yet. Thanks for bringing it to my attention!

**Comment by ryanguill** [comment]
Looks to be the related ticket: https://tracker.adobe.com/#/view/cf-4136028

**Comment by atuttle** [comment]
I think I've got it fixed. Working for me. Please verify, @anteroortiz.

**Comment by anteroortiz** [comment]
Looks good. Thank you!

**Comment by anteroortiz** [comment]
FYI: I think Update 11 may just be bad. Testing some more, outside of Taffy, and I'm seeing a case where text being pulled from the database ("open" 4 characters) is reporting a length of 5 because the first character is some symbol with an ASC() value of 2. According to the chart on the wikipedia page, it's the "Start of Text" character.

https://en.wikipedia.org/wiki/ASCII

I'm reverting back to Update 10.

**Comment by atuttle** [comment]
@anteroortiz if you read this entire thread, you'll see that `chr(2)` is the special character that Taffy is prepending to the value when you use `encode.string()` to force it to be serialized as `"1"` and not `1` (useful in cases like zip codes where the first digit could be a zero, for example). 

The updated code removes those after serialization is complete, but only using the default serializer. If you're using a custom serializer then it won't be removed. You also don't need to use `encode.string()` for alpha values like "open" because CF will correctly serialize that to a string.

**Comment by anteroortiz** [comment]
I understand what's happening with Taffy. I was just sharing that, **outside of the scope of Taffy**, I'm also seeing issues with CF11 Update 11 and how it's dealing with text. There's no reason a query value ("home"), turned into a struct, and stored in a component should behave as "home" in Update 10 but "#Ascii 2#home" in Update 11. Again, it all works in Update 10.

Just sharing for thoroughness. While anyone reading this thread might think the issue is fixed in Taffy (and it is), you may still have issues **elsewhere** in the application because the real issue might be that CF 11 Update 11 is putting the chr(2) value in front of things for some unknown reason.

**Comment by evagoras** [comment]
Perhaps Adobe is using that same hack now, either intentionally, or as a side effect of something else they have done :)

**Comment by atuttle** [comment]
My local install is upgraded to CF11u11 and I don't see the chr(2) anywhere in a normal string response...

A sample API response:

![image](https://cloud.githubusercontent.com/assets/46990/24473454/5280567e-1497-11e7-8602-fef283eb1552.png)


And if I copy that relationship string to my [unprintables finder](http://adamtuttle.codes/unprintables/), it says there aren't any:

![image](https://cloud.githubusercontent.com/assets/46990/24473497/8139e7be-1497-11e7-9a02-b99e62aee5df.png)

**Comment by anteroortiz** [comment]
Yes, Taffy is all good. 

I found the likely source of my other issues. It's with the serialization that FW/1 uses when you use it to render JSON data. It's obviously not doing the fix that Taffy applied.

**Comment by atuttle** [comment]
If someone wanted to submit a PR to https://github.com/atuttle/TaffyDocs that documents the change that resulted from this issue, that would be rad. 😎

**Comment by ryanguill** [comment]
https://github.com/atuttle/TaffyDocs/pull/17
